zebra-network/src/address_book.rs

```diff
@@ -126,34 +133,36 @@ impl AddressBook {
     }
 
     /// Get the local listener address.
-    pub fn get_local_listener(&self) -> MetaAddrChange {
+    ///
+    /// This address contains minimal state, but it is not sanitized.
+    pub fn get_local_listener(&self) -> MetaAddr {
         MetaAddr::new_local_listener(&self.local_listener)
+            .into_new_meta_addr()
+            .expect("unexpected invalid new local listener addr")
     }
 
     /// Get the contents of `self` in random order with sanitized timestamps.
     pub fn sanitized(&self) -> Vec<MetaAddr> {
         use rand::seq::SliceRandom;
         let _guard = self.span.enter();
-        let mut peers = self
-            .peers()
-            .filter_map(|a| MetaAddr::sanitize(&a))
+
+        let mut peers = self.by_addr.clone();
+
+        // Unconditionally add our local listener address to the advertised peers,
+        // to replace any self-connection failures. The address book and change
+        // constructors make sure that the SocketAddr is canonical.
+        let local_listener = self.get_local_listener();
+        peers.insert(local_listener.addr, local_listener);
```

Comment:
This change makes sure our local listener address overrides any existing entry in the address book.

But we never add our local listener to our own address book - we want to keep any failed self-connection state.